File: CS/08_Bookmarks/ExpandBookmarks.cs
using Spire.Pdf;
using Spire.Pdf.Bookmarks;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace ExpandBookmarks
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            // Create a new PDF document
            PdfDocument doc = new PdfDocument();

            // Load the file from disk
            doc.LoadFromFile(@"..\..\..\..\..\..\Data\Template_Pdf_1.pdf");

            // Set true to expand the bookmarks
            ForeachBookmark(doc.Bookmarks, true);
            String result = "ExpandBookmarks_out.pdf";

            // Save the document
            doc.SaveToFile(result);

            //Launch the Pdf file
            PDFDocumentViewer(result);
        }
        // Function to iterate through the bookmarks and expand or collapse them based on the specified flag
        private void ForeachBookmark(PdfBookmarkCollection collections, bool expand)
        {
            // Check if the collection is empty, and return if it is
            if (collections.Count == 0)
                return;

            // Iterate through each bookmark in the collection
            foreach (PdfBookmark bookmark in collections)
            {
                // Check if the current bookmark has child bookmarks
                if ((bookmark as PdfBookmarkCollection).Count > 0)
                {
                    // Recursively call the function to process child bookmarks
                    ForeachBookmark(bookmark as PdfBookmarkCollection, expand);

                    // Set the ExpandBookmark property of the current bookmark to the specified flag
                    bookmark.ExpandBookmark = expand;
                }
            }
        }

        private void PDFDocumentViewer(string filename)
        {
            try
            {
                System.Diagnostics.Process.Start(filename);
            }
            catch { }
        }
    }
}

File: VB/08_Bookmarks/ExpandBookmarks.vb
Imports Spire.Pdf
Imports Spire.Pdf.Bookmarks

Namespace ExpandBookmarks
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
			' Create a new PdfDocument instance
			Dim doc As New PdfDocument()

			' Load an existing PDF document from the specified file path
			doc.LoadFromFile("..\..\..\..\..\..\Data\Template_Pdf_1.pdf")

			' Recursively expand all bookmarks in the document
			ForeachBookmark(doc.Bookmarks, True)

			' Specify the output file name
			Dim result As String = "ExpandBookmarks_out.pdf"

			' Save the modified document to the specified file path
			doc.SaveToFile(result)

			' Close the document
			doc.Close()

			' Launch the Pdf file
			PDFDocumentViewer(result)
		End Sub
		Private Sub ForeachBookmark(ByVal collections As PdfBookmarkCollection, ByVal expand As Boolean)
			' Check if the collections variable has no elements
			If collections.Count = 0 Then
				' Exit the method if there are no collections
				Return
			End If
			' Iterate through each bookmark in the collections variable
			For Each bookmark As PdfBookmark In collections
				' Check if the current bookmark is a collection of bookmarks
				If (TryCast(bookmark, PdfBookmarkCollection)).Count > 0 Then
					' Recursively expand the sub-bookmarks within the bookmark collection
					ForeachBookmark(TryCast(bookmark, PdfBookmarkCollection), expand)

					' Set the expand state of the current bookmark based on the expand parameter
					bookmark.ExpandBookmark = expand
				End If
			Next bookmark
		End Sub
		Private Sub PDFDocumentViewer(ByVal filename As String)
			Try
				System.Diagnostics.Process.Start(filename)
			Catch
			End Try
		End Sub
	End Class
End Namespace
